dali/kernels/erase/erase_gpu.h

```diff
@@ -0,0 +1,526 @@
+        copy = false;
+      }
+    }
+    if (copy) {
+      sample.out[offset] = sample.in[offset];
+    } else {
+      sample.out[offset] = fill_value;
+    }
+  }
+};
+
+template <int channel_dim = -1, typename T, int ndim = 2>
+__global__ void erase_gpu_impl(erase_sample_desc<T, ndim> *samples, ivec<ndim> region_dims,
+                               int erase_regions_count, span<T> fill_values) {
+  const auto region_idx = blockIdx.x;
+  const auto sample_idx = blockIdx.y;
+
+  const auto &sample = samples[sample_idx];
+  ivec<ndim> sample_dims = sample.sample_shape;
+
+  auto region_box = get_region(region_idx, region_dims, sample_dims);
+
+  const auto *erase_regions_ptrs = sample.erase_regions;
+  constexpr int max_regions = 1024;
+  __shared__ ibox<ndim> erase_regions[max_regions];
+  __shared__ int filtered_region_idx;
+
+  filtered_region_idx = 0;
+  __syncthreads();
+  int full_cover = false;
+
+  // check which erase_regions overlap with current region_box
+  for (int i = threadIdx.y * blockDim.x + threadIdx.x; i < erase_regions_count;
+       i += blockDim.y * blockDim.x) {
+    if (erase_regions_ptrs[i].overlaps(region_box)) {
+      erase_regions[atomicAdd(&filtered_region_idx, 1)] = erase_regions_ptrs[i];
+      if (erase_regions_ptrs[i].contains(region_box)) {
+        full_cover = true;
+        break;
+      }
+    }
+  }
+  __syncthreads();
+
+  //  Check if we have full or no cover at all
+  int total_erase_regions = filtered_region_idx;
+  __syncthreads();
+  int is_fully_erased = __any_sync(FULL_MASK, full_cover);
+
+  if (total_erase_regions == 0) {
+    // do a full copy
+    erase_generic<do_copy, channel_dim>(sample, region_box, fill_values);
+    return;
+  } else if (is_fully_erased) {
+    // do a total erase
+    erase_generic<do_erase, channel_dim>(sample, region_box, fill_values,
+                                         make_span(erase_regions, total_erase_regions));
+    return;
+  } else {
+    erase_generic<do_copy_or_erase, channel_dim>(sample, region_box, fill_values,
+                                                 make_span(erase_regions, total_erase_regions));
+  }
+}
+
+template <int ndim>
+ivec<ndim> to_ivec(TensorShape<ndim> shape) {
+  ivec<ndim> result;
+  for (int d = 0; d < ndim; d++) {
+    result[d] = shape[d];
+  }
+  return result;
+}
+
+template <typename T, int ndim, int channel_dim = -1>
+struct EraseGpu {
+  static_assert(
+      -1 <= channel_dim && channel_dim < ndim,
+      "Channel dim can either be ignored (for channel_dim = -1) or must be in [0, ndim) interval.");
+  using sample_t = erase_sample_desc<T, ndim>;
+
+  KernelRequirements Setup(KernelContext &context,
+                          //  const TensorListView<StorageGPU, const T, ndim> &in
+                           const InListGPU<T, ndim> &in,
+                           const InListGPU<ibox<ndim>, 1> &erased_regions,
+                           const std::vector<T> &fill_values = {0}) {
```

Comment:
span maybe?

Reply:
either that, or `InTensorCPU<T, 1>`

Reply:
On top of that - I think we can safely say that the requirement of passing everything to Setup _and_ Run was proven unnecessary - so you can just leave out if it's not used before Run.

Reply:
As you can see, I'm doing error checking in setup based on the size of fill_values.

Reply:
And I don't think there is a nice way of providing default value in case of span, so unless someone suggests something, I would opt for vector. We cannot create span from rvalue, the view also wouldn't work without some additional backing memory unless I'm wrong.

Reply:
Done by specialcasing empty span